File: NC-150/06_Linked_List/36_merge_two_sorted_lists.py
"""
Problem: Merge Two Sorted Lists (LeetCode 21)

Merge two sorted linked lists and return it as a sorted list.
The list should be made by splicing together the nodes of the first two lists.

Time Complexity: O(n + m) where n, m are lengths of the two lists
Space Complexity: O(1) for iterative, O(n + m) for recursive
"""

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        """Iterative approach with dummy node - O(1) space"""
        dummy = ListNode(0)
        tail = dummy
        
        while list1 and list2:
            if list1.val <= list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next
        
        # Attach remaining nodes
        tail.next = list1 if list1 else list2
        
        return dummy.next
    
    def mergeTwoListsRecursive(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        """Recursive approach - O(n+m) space due to call stack"""
        # Base cases
        if not list1:
            return list2
        if not list2:
            return list1
        
        # Choose smaller value as head, recurse for the rest
        if list1.val <= list2.val:
            list1.next = self.mergeTwoListsRecursive(list1.next, list2)
            return list1
        else:
            list2.next = self.mergeTwoListsRecursive(list1, list2.next)
            return list2

# Helper functions
def create_list(values):
    if not values:
        return None
    head = ListNode(values[0])
    current = head
    for val in values[1:]:
        current.next = ListNode(val)
        current = current.next
    return head

def list_to_array(head):
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result

if __name__ == "__main__":
    sol = Solution()
    
    # Test case 1: [1,2,4] + [1,3,4] = [1,1,2,3,4,4]
    l1 = create_list([1, 2, 4])
    l2 = create_list([1, 3, 4])
    merged1 = sol.mergeTwoLists(l1, l2)
    print(f"Merged: {list_to_array(merged1)}")
    assert list_to_array(merged1) == [1, 1, 2, 3, 4, 4]
    
    # Test case 2: [] + [] = []
    l3 = create_list([])
    l4 = create_list([])
    merged2 = sol.mergeTwoLists(l3, l4)
    print(f"Merged: {list_to_array(merged2)}")
    assert list_to_array(merged2) == []
    
    # Test case 3: [] + [0] = [0]
    l5 = create_list([])
    l6 = create_list([0])
    merged3 = sol.mergeTwoLists(l5, l6)
    print(f"Merged: {list_to_array(merged3)}")
    assert list_to_array(merged3) == [0]
    
    # Test case 4: Different lengths
    l7 = create_list([1, 3, 5, 7])
    l8 = create_list([2, 4])
    merged4 = sol.mergeTwoLists(l7, l8)
    print(f"Merged: {list_to_array(merged4)}")
    assert list_to_array(merged4) == [1, 2, 3, 4, 5, 7]
    
    # Test recursive solution
    l9 = create_list([1, 2, 4])
    l10 = create_list([1, 3, 4])
    merged5 = sol.mergeTwoListsRecursive(l9, l10)
    print(f"Recursive merged: {list_to_array(merged5)}")
    assert list_to_array(merged5) == [1, 1, 2, 3, 4, 4]
    
    print("\nAll test cases passed!")

File: NC-150/06_Linked_List/36_merge_two_sorted_lists.cpp
/*
 * Problem: Merge Two Sorted Lists (LeetCode 21)
 * 
 * Merge two sorted linked lists and return it as a sorted list.
 * The list should be made by splicing together the nodes of the first two lists.
 * 
 * Time Complexity: O(n + m) where n, m are lengths of the two lists
 * Space Complexity: O(1) for iterative, O(n + m) for recursive
 */

#include <iostream>
#include "list_node.h"

class Solution {
public:
    // Solution 1: Iterative with dummy node
    // Time: O(n + m), Space: O(1)
    ListNode* mergeTwoListsIterative(ListNode* list1, ListNode* list2) {
        // Dummy node simplifies edge cases
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (list1 != nullptr && list2 != nullptr) {
            if (list1->val <= list2->val) {
                tail->next = list1;
                list1 = list1->next;
            } else {
                tail->next = list2;
                list2 = list2->next;
            }
            tail = tail->next;
        }
        
        // Attach the remaining list (one of them is empty)
        tail->next = (list1 != nullptr) ? list1 : list2;
        
        return dummy.next;
    }
    
    // Solution 2: Recursive
    // Time: O(n + m), Space: O(n + m) for recursion stack
    ListNode* mergeTwoListsRecursive(ListNode* list1, ListNode* list2) {
        // Base cases
        if (list1 == nullptr) return list2;
        if (list2 == nullptr) return list1;
        
        // Choose smaller value as head, recurse for the rest
        if (list1->val <= list2->val) {
            list1->next = mergeTwoListsRecursive(list1->next, list2);
            return list1;
        } else {
            list2->next = mergeTwoListsRecursive(list1, list2->next);
            return list2;
        }
    }
    
    // Main function using iterative approach
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        return mergeTwoListsIterative(list1, list2);
    }
};

int main() {
    Solution sol;
    
    // Test case 1: [1,2,4] + [1,3,4] = [1,1,2,3,4,4]
    ListNode* l1 = createList({1, 2, 4});
    ListNode* l2 = createList({1, 3, 4});
    std::cout << "List 1: ";
    printList(l1);
    std::cout << "List 2: ";
    printList(l2);
    ListNode* merged1 = sol.mergeTwoLists(l1, l2);
    std::cout << "Merged: ";
    printList(merged1);
    deleteList(merged1);
    
    // Test case 2: [] + [] = []
    ListNode* l3 = createList({});
    ListNode* l4 = createList({});
    std::cout << "\nList 1: ";
    printList(l3);
    std::cout << "List 2: ";
    printList(l4);
    ListNode* merged2 = sol.mergeTwoLists(l3, l4);
    std::cout << "Merged: ";
    printList(merged2);
    deleteList(merged2);
    
    // Test case 3: [] + [0] = [0]
    ListNode* l5 = createList({});
    ListNode* l6 = createList({0});
    std::cout << "\nList 1: ";
    printList(l5);
    std::cout << "List 2: ";
    printList(l6);
    ListNode* merged3 = sol.mergeTwoLists(l5, l6);
    std::cout << "Merged: ";
    printList(merged3);
    deleteList(merged3);
    
    // Test case 4: Different lengths [1,3,5,7] + [2,4]
    ListNode* l7 = createList({1, 3, 5, 7});
    ListNode* l8 = createList({2, 4});
    std::cout << "\nList 1: ";
    printList(l7);
    std::cout << "List 2: ";
    printList(l8);
    ListNode* merged4 = sol.mergeTwoLists(l7, l8);
    std::cout << "Merged: ";
    printList(merged4);
    deleteList(merged4);
    
    std::cout << "\nAll test cases passed!" << std::endl;
    return 0;
}

